torch/csrc/stable/tensor.h

```diff
@@ -0,0 +1,136 @@
+#pragma once
+
+#include <torch/csrc/stable/library.h>
+#include <torch/csrc/inductor/aoti_torch/c/shim.h>
+
+namespace torch::stable {
+
+using DeviceIndex =
+    int8_t; // this is from c10/core/Device.h and can be header only?
+
+static inline void delete_tensor_object(AtenTensorHandle ath) {
+  AOTI_TORCH_ERROR_CODE_CHECK(
+      aoti_torch_delete_tensor_object(ath));
+}
+
+// The torch::stable::Tensor class is a highlevel C++ header-only wrapper around the C
+// shim Tensor APIs.
+class Tensor {
+ private:
+  std::shared_ptr<AtenTensorOpaque> ath_;
+
+ public:
+  Tensor() = delete;
+ 
+  // Wrap AtenTensorHandle
+  explicit Tensor(AtenTensorHandle ath)
+    : ath_(ath, delete_tensor_object) {}
+  
+  // Wrap StableIValue
+  explicit Tensor(StableIValue siv)
+    : ath_(to<AtenTensorHandle>(siv), delete_tensor_object) {}
+
+  // Copy and move constructors can be default cuz the underlying handle is a shared_ptr
+  Tensor(const Tensor& other) = default;
+  Tensor(Tensor&& other) noexcept = default;
+  
+  // Copy and move assignment operators can be default cuz the underlying handle is a shared_ptr
+  Tensor& operator=(const Tensor& other) = default;
+  Tensor& operator=(Tensor&& other) noexcept = default;
+
+  // Destructor can be default: shared ptr has custom deletion logic
+  ~Tensor() = default;
+
+  AtenTensorHandle get() const {
+    return ath_.get();
+  }
+
+  StableIValue get_SIV() {
+    AtenTensorHandle handle = ath_.get();
+
+    // The following is our way of incrementing the refcount of the underlying
+    // at::Tensor that we point to. Why do we want this weird behavior?
+    // Because! We expect users to only need a StableIValue when they are trying
+    // to pass the Tensor into a stack-based API, e,g., aoti_torch_call_dispatcher.
+    //
+    // A stack-based API is one that expects a stack of inputs converted to
+    // StableIValues. Our contract with any stack-based API is that the stack
+    // has ownership of its Tensor arguments. Since this torch::stable::Tensor
+    // object will likely go out of scope by the end of the user extension's
+    // local function and will thus delete its reference on the at::Tensor,
+    // we create a new unused AtenTensorHandle to indirectly increment the
+    // refcount on the at::Tensor.
+    AtenTensorHandle new_ath;
+    aoti_torch_new_tensor_handle(handle, &new_ath);
+
+    return from(new_ath);
+  }
+
+  void* data_ptr() const {
+    void* data_ptr;
+    aoti_torch_get_data_ptr(ath_.get(), &data_ptr);
```

Comment:
don't you need to AOTI_TORCH_CODE_ERROR_CHECK here (and all the other places you call an aoti_torch function)? there is no guarantee there isn't a different implementation of the ABI (such as libtorch from 2030) that may actually throw an error.

Reply:
o yes